### DISCUSSION TITLE
Request to make NullPool the default - fixes massive footgun with Render.com and DigitalOcean

### DISCUSSION THREAD

**Discussion by zackees** [sqlalchemy/sqlalchemy#10238] [Usage Questions]
### Describe the bug

Hi there, I found a nasty bug that is affecting literally ALL of our projects for the org that I work for. I did it, the contractor that work for us did it. So I wanted to bring this up for discussion.

# SQL Alchemy uses a connection Pool by default

So what's the big deal? Well, both DigitalOcean and Render.com kill these connections at the os level after a certain amount of time has passed. SQLAlchemy then *blindly* attempts to use these dead connections and bombs out the server request with an error 500. My expectation as a developer would be that SQLAlchemy could auto-recover from this but it doesn't, and this is surprising.

# This bug is nasty and sporadic and infrequently happens during development

During an app redeploy, all connections are cleared out and the pool is torn down and then reinstantiated when the app is brought back online. As a result, these dead connections are largely hidden from the developer and go away as soon as they start pushing code, for example adding logging to investigate. Then after a length of time the app starts to degrade again. I caught this issue because I added logging in the proper place and was able to see the server logs to see the error:

```
(psycopg2.OperationalError) SSL connection has been closed unexpectedly
```

# New apps don't need a connection pool until they start to scale

And when apps need to scale, developers are willing to put in the time and energy to find these optimizations. In my opinion, a connection pool should be thought of in 2023 as **dangerous**, due to the number of Docker hosts coming online that auto-kill these connections after not too much time. Maybe a decade ago everyone had a default DB config with long running connections that could remain open, but this certainly isn't the case now.

# Does adding pre_ping=True as the default solve the issue?

Mostly, but I'm still getting 500 errors. I have to use very aggressive settings to make them go away. This is what I found works:

```python
engine = create_engine(
    DB_URL,
    pool_pre_ping=True,
    pool_recycle=60,
    pool_size=10,
    max_overflow=2,
    pool_use_lifo=True,
)
```

However, the app is early stage so I just opted for NullPool instead because I don't need the extra 50 ms or so that a connection pool gives. And when I re-introduce a connection pool in the future - I'll thoroughly test it.

# I can only imagine how many people this bug is affecting

Given that both I and a contractor made this mistake in *every* backend microservice (we have about 6), I'm going to go ahead and speculate that this is happening to **most** of the people using sqlalchemy for Render.com/DigitalOcean. That's a *lot* of broken code out there, right at the beginning of an app when the developer is in the learning curve. This is nasty footgun.

I urge the developers to change the defaults somehow so that sqlalchemy works out of the box. It's been pretty amazing for us to use this tool and we applaud all the hard work that has been put into this middlewear. But wow - this one was a doozy.

### SQLAlchemy Version in Use

2.0.17

### DBAPI (i.e. the database driver)

psycopg2 (but I think but can't confirm it's happening in mysql)

### Database Vendor and Major Version

Latest PostgreSQL from Render.com

### Python Version

python 3.10, 3.11 (probably exists in all versions)

### Operating system

Ubuntu, Alpine

### To Reproduce

```python
Create a PostgreSQL db on Render.com.

Make a fastapi python project with a simple table.

Have the front end python app do a request on the DB. Doesn't matter what it is, it can be anything.

Launch the CRUD app.

Let it run for a while.

Randomly hit it with a request. You'll see a 500 error that occurs. If you log it you'll see:


(psycopg2.OperationalError) SSL connection has been closed unexpectedly
```

The exception will be caught in a default handler and the request will return HTTP error code 500.

Now change the pool class to NullPool and re-deploy.

The error will now go away.
```


### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
``` 

### Additional context

_No response_

**Comment by zzzeek** [comment]
hi - 

I've converted this to a discussion, so we can figure out more specifically what problem you are having.  there is no "bug" in the connection pool, it does what it is designed to do as it has been for literally 18 years, which is to pool connections so that they are available for fast re-use.    For database server connections that have a server-side timeout, the [pool_recycle](https://docs.sqlalchemy.org/en/20/core/pooling.html#setting-pool-recycle) setting, which prevents the use of a stale connection, should be used, which again is an option that has been in widespread use for such backends as MySQL since SQLAlchemy's first releases in 2006.

> So what's the big deal? Well, both DigitalOcean and Render.com kill these connections at the os level after a certain amount of time has passed.

Right, so take that "certain amount of time", and then set a value of about half of that "time" as the `pool_recycle` setting.  this will cause any connection more than N seconds old to be immediately discarded when checked out and replaced.   

that's really all that seems to be happening here, so hopefully that setting is all you need.

**Reply by zzzeek** [reply]
I would also add that pre_ping should fix this also, **unless** the specific error being raised is not being matched as a "disconnect" error, or if the error is being raised mid-transaction, which the pool choice would not have any ability to affect.   I would need to see a stack trace to confirm. However, the error "SSL connection has been closed unexpectedly" is well known and [is matched](https://github.com/sqlalchemy/sqlalchemy/blob/e5a81c980b5c07f7d980b90149c6b18ea16f6f4b/lib/sqlalchemy/dialects/postgresql/psycopg2.py#L863) by current disconnect logic.  So if the connection is server-closed and that's the error it raises on ping, pre-ping would recycle this connection.

if you have a stack trace that illustrates the connection in some other kind of "closed" state for which pre-ping is not detecting, we can open a separate issue for that but only with specific error message text and complete stack trace confirming pre-ping is erroring out.

**Reply by zackees** [reply]
I still saw errors with a 60 second timeout though. It just made the problem from somewhat infrequent, to very infrequent.

Connection pools are an advanced late stage performance optimization. I'm actually very surprised that a connection pool is the default at all, but maybe this made sense 18 years ago.

Also, I'm just trying to point out this foot gun. You don't have any obligation to fix it, but know that it's affecting a large amount of developers.

**Reply by zzzeek** [reply]
> due to the number of Docker hosts coming online that auto-kill these connections after not too much time. 

that would suggest the connection pool would ship with a default pool_timeout setting that is lower than the general amount used by these "Docker hosts", which I'm generally not familiar with.    

> I still saw errors with a 60 second timeout though. It just made the problem from somewhat infrequent, to very infrequent.

that means something else is still going on.

> Connection pools are an advanced late stage performance optimization. I'm actually very surprised that a connection pool is the default at all, but maybe this made sense 18 years ago.

please try to consider the practicality of bluntly switching off this behavior for the millions of servers upon which SQLAlchemy is currently running, which at the very least would cause people's pool settings like "pool_size" to immediately raise errors.      It's not unusual that managed database services (which I dont think are the most common default at all) would have server-side timeouts and that developers need to be aware of these when using such managed services.

> Also, I'm just trying to point out this foot gun. 

I don't even see this as a "footgun" - a footgun refers to a library whose structure and design lends itself towards developers writing code that is non-obviously incorrect.  this is not a coding mistake, just a simple configuration option that would be made at the same time one is reading the documentation for how to connect to a particular cloud-managed database, where that cloud service should also be indicating **in the same document** if they have server-side connection timeouts implemented as well.

**Reply by CaselIT** [reply]
The main issue I see here is on Render.com/DigitalOcean side.
They should not close connections automatically after that should of a time.

Since I'm assuming you are paying them for the privilege of using their services I would raise the issue to them

**Reply by CaselIT** [reply]
just pointing out that if the timeout is less than 60s, queries that take that amount of time are not unheard of, especially in some analytics context, so that would mean that such a service would be completely unusable on that cloud provider

**Comment by CaselIT** [comment]
Hi,

I'm -1 on this. You seem to make a big deal of a deficiency in your deployment environment. My suggestion would to use something else

> And when apps need to scale, developers are willing to put in the time and energy to find these optimizations. In my opinion, a connection pool should be thought of in 2023 as dangerous, due to the number of Docker hosts coming online that auto-kill these connections after not too much time. Maybe a decade ago everyone had a default DB config with long running connections that could remain open, but this certainly isn't the case now.

I'm not convinced that docker or k8s pose a problem here. I've used sqlalchemy in multiple clouds on docker / k8s without similar issues

> I can only imagine how many people this bug is affecting

given by the number of reports we get, I would guess very little.

> Given that both I and a contractor made this mistake in every backend microservice (we have about 6), I'm going to go ahead and speculate that this is happening to most of the people using sqlalchemy for Render.com/DigitalOcean. That's a lot of broken code out there, right at the beginning of an app when the developer is in the learning curve. This is nasty footgun.

The only thing this tells me is not to use those services. I don't want to use a service that randomly closes a connection after a very short while if I don't tell it to.

**Reply by zackees** [reply]
> given by the number of reports we get, I would guess very little.

Okay, well you would know better than I.

Thanks for your attention to this and I'm happy to have gotten this on your radar.

**Comment by Abdur-rahmaanJ** [comment]
Idk what the package did but not i need to set NullPool or engine.dispose. A true 2.0.

**Reply by CaselIT** [reply]
Nothing has changed between 1.4 and 2.0 regarding pool on any database other than SQLite. In v2 sqlite also defaults to queue pool.

I would suggest opening another discussion if you are not using sqlite

**Reply by Abdur-rahmaanJ** [reply]
Did not know that! Using SQLite.

**Reply by CaselIT** [reply]
changelog here https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-aba109875059299e0eb82359b5825416
